jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -191,40 +191,64 @@ public int getMaxCapacity()
     public RetainableByteBuffer acquire(int size, boolean direct)
     {
         RetainedBucket bucket = bucketFor(size, direct);
+
+        // No bucket, return non-pooled.
         if (bucket == null)
-            return newRetainableByteBuffer(size, direct, this::removed);
+            return newRetainableByteBuffer(size, direct, null);
+
+        // Try to acquire a pooled entry.
         Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
+        if (entry != null)
+        {
+            RetainableByteBuffer buffer = entry.getPooled();
+            ((Buffer)buffer).acquire();
+            return buffer;
+        }
 
-        RetainableByteBuffer buffer;
+        return newRetainableByteBuffer(bucket.getCapacity(), direct, buffer -> reserve(bucket, buffer, direct));
+    }
+
+    private void reserve(RetainedBucket bucket, RetainableByteBuffer buffer, boolean direct)
+    {
+        // Discard the buffer if maxMemory is exceeded.
+        if (isMaxMemoryExceeded(bucket, direct))
+            return;
+
+        Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().reserve();
+        // Cannot reserve, discard the buffer.
         if (entry == null)
-        {
-            Pool.Entry<RetainableByteBuffer> reservedEntry = bucket.getPool().reserve();
-            if (reservedEntry != null)
-            {
-                buffer = newRetainableByteBuffer(bucket._capacity, direct, retainedBuffer ->
-                {
-                    BufferUtil.reset(retainedBuffer.getByteBuffer());
-                    if (!reservedEntry.release())
-                        reservedEntry.remove();
-                });
+            return;
+
+        ByteBuffer byteBuffer = buffer.getByteBuffer();
+        BufferUtil.reset(byteBuffer);
+        Buffer pooledBuffer = new Buffer(byteBuffer, b -> release(bucket, entry, direct));
+        // Discard the buffer if the entry cannot be enabled.
+        if (entry.enable(pooledBuffer, false))
+            updateMemory(bucket.getCapacity(), direct);
+        else
+            entry.remove();
+    }
 
-                // A reserved entry may become old and be evicted before it is enabled.
-                if (reservedEntry.enable(buffer, true))
-                {
-                    if (direct)
-                        _currentDirectMemory.addAndGet(buffer.capacity());
-                    else
-                        _currentHeapMemory.addAndGet(buffer.capacity());
-                    releaseExcessMemory(direct);
-                    return buffer;
-                }
-            }
-            return newRetainableByteBuffer(size, direct, this::removed);
+    private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> entry, boolean direct)
+    {
+        BufferUtil.reset(entry.getPooled().getByteBuffer());
+        if (isMaxMemoryExceeded(bucket, direct))
+        {
+            bucket.evict(entry);
         }
```

Comment:
Can this be true often?   The `reserve` method will not add a buffer if that will exceed the max memory, so for the vast majority of releases, the maxMemoryExceeded check is wasted work.    You might as well do the check in `reserve` to not add the buffer if it would exceed, and then you never need evict.